RapidPro SMS gateway documentation (#462)
medic/cht-docs: content/en/apps/guides/messaging/rapidpro.md

---
title: RapidPro Messaging Gateway
linkTitle: "RapidPro"
weight:
description: >
    Integration for sending and receiving messages
relatedContent: >
    apps/reference/app-settings/sms
    apps/guides/forms/app-form-sms
    apps/guides/messaging/sms-states
    apps/guides/messaging/message-loops
    apps/guides/messaging/shortcodes
---

As of v3.11.0, messages can be sent and received using [RapidPro]({{% ref "apps/features/integrations/rapidpro" %}}) as a messaging gateway. 

## RapidPro configuration

### Store globals

Generate a long unique key to use as the `cht_api_key`.

Log in to your RapidPro dashboard, go to the globals page (`/global/`) and create two globals with the following data:

> - name: `cht_url`, value: `https://<your-cht-instance-host>/api/v1/sms/radpidpro/incoming-messages`. For security the instance host **must not** include basic authentication.

Reviewer: Consider

For security,.....

> ---
> title: RapidPro SMS Gateway
> linkTitle: "RapidPro"
> weight:
> description: >
>     Integration for sending and receiving SMS
> relatedContent: >
>     apps/reference/app-settings/sms
>     apps/guides/forms/app-form-sms
>     apps/guides/messaging/sms-states
>     apps/guides/messaging/message-loops
>     apps/guides/messaging/shortcodes
> ---

Reviewer 2: Similar to [the other comment](https://github.com/medic/cht-docs/pull/462/files#r608814872), RapidPro can be used for messaging beyond SMS too so the title should reflect that too

```suggestion
---
title: RapidPro Messaging Gateway
linkTitle: "RapidPro"
weight:
description: >
    Integration for sending and receiving messages
relatedContent: >
    apps/reference/app-settings/sms
    apps/guides/forms/app-form-sms
    apps/guides/messaging/sms-states
    apps/guides/messaging/message-loops
    apps/guides/messaging/shortcodes
---
```

As of v3.11.0, SMS messages can be sent and received using [RapidPro](https://community.rapidpro.io/).

## RapidPro configuration

> 

Reviewer 3: I think the first question we'll get is "how do I access rapidpro?". I think it'd be worth adding something about step 1 being contact Medic to host, or use TextIt, or host your own.

Author: There's already a section about RapidPro here: https://docs.communityhealthtoolkit.org/apps/features/integrations/rapidpro/

Would it suite to link to that and reduce duplicate content?

Reviewer 3: Yes, linking is perfect, thanks.

> As of v3.11.0, SMS messages can be sent and received using [RapidPro]({{% ref "apps/features/integrations/rapidpro" %}}) as an SMS Gateway. 

Reviewer 2: It's probably best to be more generic about the message type, since outgoing messages could be sent via channels other than SMS (eg Telegram).

```suggestion
As of v3.11.0, messages can be sent and received using [RapidPro]({{% ref "apps/features/integrations/rapidpro" %}}) as an messaging gateway. 
```

### Store globals

> Generate a long unique key to use as the `medic-api-key`.

Reviewer 4: Is there a reason we call this `medic-api-key` instead of `cht-api-key`?

Author: Either one works, the name is irrelevant. just a convention to be used in this doc. We can also call it "bubblegum_jewel" :D

Author: The sole reason for `medic-api-key` is that it's the same name used in the Africa's Talking documentation (and it was my 
"model")

Reviewer 4: Got it, thanks. If it's just a couple minutes to change in both places, it's probably worth it.  If it takes us down a deep rabbit hole, probably not worth it right now.

Reviewer 4: Thanks for the switch!

> Log in to your RapidPro dashboard, go to the globals page (`/globals/`) and create two globals with the following data:

Reviewer 5: on textit this is `/global/`

Author: Oops, sorry, I'll fix it.

Author: Fixed!

> - name: `cht_url`, value: `https://<your-cht-instance-host>/api/v1/sms/radpidpro/incoming-messages`. The instance host should not include basic authentication.

Reviewer 3: I assume we don't want to include basic auth because then the admin password will be stored in plaintext in rapidpro somewhere. If so, I think we can use stronger wording here.

```suggestion
- name: `cht_url`, value: `https://<your-cht-instance-host>/api/v1/sms/radpidpro/incoming-messages`. For security the instance host **must** not include basic authentication.
```

- name: `cht_api_key`, value: `<cht_api_key>`

> The names of these two globals are flexible, but in this document we will keep referring at the names defined above. 

Reviewer 3: ```suggestion
The names of these two global variables are arbitrary, but in this document we will keep referring at the names defined above. 
```

Author: Done.

## CHT Core configuration

### API keys

The API keys should be treated as securely as a password as anyone with access to them will be able to send messages using your account. Therefore, we store them in the CouchDB configuration.

To add the credentials to the admin config you need to either [PUT the value using curl](https://docs.couchdb.org/en/stable/api/server/configuration.html#put--_node-node-name-_config-section-key) or similar:

```sh
curl -X PUT https://<user>:<pass>@<domain>/_node/couchdb@127.0.0.1/_config/medic-credentials/rapidpro:incoming -d '"<cht_api_key>"'
curl -X PUT https://<user>:<pass>@<domain>/_node/couchdb@127.0.0.1/_config/medic-credentials/rapidpro:outgoing -d '"<rapidpro_api_key>"'
```

{{% alert title="Note" %}}
`couchdb@127.0.0.1` is the local node name, and may be different for you depending on your setup.
{{% /alert %}}

You can also add it via Fauxton:
- Navigate to [the Config screen](http://localhost:5984/_utils/#/_config)
- Click `Add Option`
- The `Section` should be `medic-credentials`, the `Name` should be `rapidpro:incoming` or `rapidpro:outgoing`, and the value should be the relevant API key.
- Click `Create`

> - You should then be able to see your credential in the list of configuration shown

Reviewer 3: This section is basically the same as Outbound and Africa's Talking pages. We should create a new page for "credentials" and then just link to it from the three pages. I don't want to distract you from RapidPro, so feel free to raise an issue to do it later.

Author: Added issue: https://github.com/medic/cht-docs/issues/463

### App settings

Update your app settings as follows.

```json
{
  "sms": {
    "outgoing_service": "rapidpro",
    "rapidpro": {
      "url": "<RapidPro instance url>"
    }
  }
}
```

{{% alert title="Rate limiting" %}}
The RapidPro API endpoints, which this integration relies on, are rate-limited. The number of messages that can be sent and the number of status updates that can be requested is limited. Check [RapidPro API reference](https://rapidpro.io/api/v2/) for more details.

If an action (sending a message or retrieving a status update) fails, it will be retried again later.
{{% /alert %}}

{{% alert title="Caveat when switching outgoing message services" %}}

> When switching from another outgoing message service to RapidPro, or switching between RapidPro accounts or hosts, make sure that none of the outgoing messages on your instance, that have been relayed through the previous service, are in a non-final state ("received-by-gateway", "forwarded-by-gateway" or "sent"). CHT Core will request state updates for these messages from RapidPro, fail on every retry, and consume your request quota.  

Reviewer 3: Slight rewording for readability...

```suggestion
When switching from another outgoing message service to RapidPro, or switching between RapidPro accounts or hosts, make sure that all of the outgoing messages on your instance, that have been relayed through the previous service, are in a final state ("delivered" or "failed"). CHT Core will request state updates for non-final messages from RapidPro, which will fail on every retry and consume your request quota.  
```

Author: This is how I worded it initially, actually. But I changed my mind when I had to list all the other "final" state we have on the CHT: "muted", "denied", "duplicate", "cleared". I should list all, correct? These messages are not used by RapidPro, but could be used by other services.

Reviewer 3: Ahh right. Yeah that's a long list. I'll leave it to you to either ignore this suggestion or list all final states - whichever you think is best.

Author: I've listed all final states :)

> {{% alert title="Rate limiting" %}}
> The RapidPro API endpoints, which this integration relies on, are rate-limited. The number of messages that can be sent and the number of status updates that can be requested is limited. Check [RapidPro API reference](https://rapidpro.io/api/v2/) for more details.
> 
> If an action (sending a message or retrieving a status update) fails, it will be retried again later.
> {{% /alert %}}
> 
> {{% alert title="Caveat when switching outgoing message services" %}}
> When switching from another outgoing message service to RapidPro, or switching between RapidPro accounts or hosts, make sure that all of the outgoing messages on your instance, that have been relayed through the previous service, are in a final state ("delivered", "failed", "denied", "cleared"). CHT Core will request state updates for non-final messages from RapidPro, which will fail on every retry and consume your request quota.
> {{% /alert %}}

Reviewer 2: I had a bit of trouble wrapping my head around these notes. Given that they seem related I've attempted to rewrite it based on my understanding, and combined to a single note:

```suggestion
{{% alert title="Rate limiting" %}}
The RapidPro API endpoints are rate-limited, meaning that requests to send or check the status of messages beyond a certain number per hour will be blocked. The limit is currently 2500 actions per hour, and may change without notice. Check out the [RapidPro API reference](https://rapidpro.io/api/v2/#rate-limiting) for more details. If sending a message or retrieving a status update fails, it will be retried automatically again later.

When the outgoing message service is set to RapidPro or the host/account are changed, CHT Core will request state updates for all messages that aren't in one of the final states: `delivered`, `failed`, `denied`, or `cleared`. Each of these request counts towards your quota. Since the messages are unlikely to exist on the new RapidPro service, these requests will fail on every retry and consume your request quota. It is therefore important that all outgoing messages have a final state before switching RapidPro accounts or hosts. The status can be set to `failed` for messages that should not be resent without user intervention, or to `scheduled` for those that should be automatically sent with the new RapidPro account.
{{% /alert %}}
```

It wasn't clear to me _how_ you'd clear or reset the status of non-final messages, so it would be nice to fill that in, eg "The status can be set {using a script} to `failed`...".

Author: Thanks for the feedback @abbyad ! Do you think we should refer to a script that doesn't exist? Should we provide this script? When discussing with @garethbowen , our conclusion was that we could provide a script when the need arises.

Reviewer 2: Sounds good. Pinging @kennsippell @derickl @michaelkohn here as a heads up since app developers may encounter this situation and not know that a script could be built for it, or how to go about doing it efficiently.

---

> As of v3.11.0, messages can be sent and received using [RapidPro]({{% ref "apps/features/integrations/rapidpro" %}}) as an messaging gateway. 

Reviewer 2: Looks good, just a remnant `n` to remove.

```suggestion
As of v3.11.0, messages can be sent and received using [RapidPro]({{% ref "apps/features/integrations/rapidpro" %}}) as a messaging gateway. 
```
